### ISSUE TITLE
AudioTagService throws InvalidOperationException

### ISSUE THREAD

**Issue by LanceMcCarthy** [Readarr/Readarr#1593]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

At the end of a scan and import cycle (manual or automatic), there is an InvalidOperationException. Looking at the logs, I have the full exception's output, which is posted below.

> I am an experienced .NET developer, so I am familiar with reading stack traces and exception messages. It looks like the problem might have happened when the tags were being read after an import.

My gut tells me this might be happening here:
https://github.com/Readarr/Readarr/blob/f55fda9bac677130ef7c5da6e2bce76b36794cc5/src/NzbDrone.Core/MediaFiles/AudioTagService.cs#L71-L78

That would explain why trying to access `.Value` on a nullable object would throw the exact exception I'm seeing if the value is null.

## Moving forward

Is there a way to get a reference to the actual `BookFile` object that caused the exception to validate my suspicion?

I'm not familiar with Readarr's code base, so take this suggestion with a grain of salt... maybe use a null accessor when attempting to access those initial values so that there's no exception thrown and the value is ignored

```csharp
var edition = trackfile?.Edition?.Value;
var book = edition?.Book?.Value;
var author = book?.Author?.Value;
var partCount = edition?.BookFiles?.Value.Count;
```

Thank you!

### Expected Behavior

The scan completes without error.

### Steps To Reproduce

Start a Refresh and Scan cycle

'Reading files' phase finishes successfully
![image](https://user-images.githubusercontent.com/3520532/159125492-d8f124df-2848-48a4-9c09-eaf3e5f90cc4.png)

'Identification' phase finishes successfully:
![image](https://user-images.githubusercontent.com/3520532/159125499-b7322e30-28b6-4ab1-a857-e2e9b7155c5a.png)

'Importing' phase finishes successfully:
![image](https://user-images.githubusercontent.com/3520532/159125533-1db03fce-fe60-476b-b647-760631c3e8aa.png)

As soon as the import finishes, I get a Failed status:
![image](https://user-images.githubusercontent.com/3520532/159125552-9aa78cdf-ffe8-4fd3-a701-bf8c6e2d91f4.png)

I know the previous items worked fine, so I opened the logs to see what exactly failed. 

### Environment

```markdown
- OS: Windows 10 (19041)
- Readarr: 0.1.0.919
- Docker Install: false
- Using Reverse Proxy: false
- Browser: n/a
```

### What branch are you running?

Master

### Trace Logs?

[edit 1] Added code block formatting to better read the exception message

```
[v0.1.0.919] System.InvalidOperationException: Nullable object must have a value.
   at NzbDrone.Core.MediaFiles.AudioTagService.GetTrackMetadata(BookFile trackfile) in D:\a\1\s\src\NzbDrone.Core\MediaFiles\AudioTagService.cs:line 73
   at NzbDrone.Core.MediaFiles.AudioTagService.WriteTags(BookFile trackfile, Boolean newDownload, Boolean force) in D:\a\1\s\src\NzbDrone.Core\MediaFiles\AudioTagService.cs:line 192
   at NzbDrone.Core.MediaFiles.BookImport.ImportApprovedBooks.Import(List`1 decisions, Boolean replaceExisting, DownloadClientItem downloadClientItem, ImportMode importMode) in D:\a\1\s\src\NzbDrone.Core\MediaFiles\BookImport\ImportApprovedBooks.cs:line 216
   at NzbDrone.Core.MediaFiles.DiskScanService.Scan(List`1 folders, FilterFilesType filter, Boolean addNewAuthors, List`1 authorIds) in D:\a\1\s\src\NzbDrone.Core\MediaFiles\DiskScanService.cs:line 170
   at NzbDrone.Core.MediaFiles.DiskScanService.Execute(RescanFoldersCommand message) in D:\a\1\s\src\NzbDrone.Core\MediaFiles\DiskScanService.cs:line 313
   at NzbDrone.Core.Messaging.Commands.CommandExecutor.ExecuteCommand[TCommand](TCommand command, CommandModel commandModel) in D:\a\1\s\src\NzbDrone.Core\Messaging\Commands\CommandExecutor.cs:line 116
   at System.Dynamic.UpdateDelegates.UpdateAndExecuteVoid3[T0,T1,T2](CallSite site, T0 arg0, T1 arg1, T2 arg2)
   at NzbDrone.Core.Messaging.Commands.CommandExecutor.ExecuteCommands() in D:\a\1\s\src\NzbDrone.Core\Messaging\Commands\CommandExecutor.cs:line 42
```

[AB#2810](https://dev.azure.com/Servarr/7ab38f4e-5a57-4d70-84f4-94dd9bc5d6df/_workitems/edit/2810)

**Comment by bakerboy448** [comment]
> 0.1.0.919

- You're a few hundred builds behind  and also there is no `master` branch.
- Upgrade to the current version and double check to see if this still persists
- just the exception is not trace logs and rarely helpful - see the gathering logs link that the bug report template linked for how to get and provide trace logs

Otherwise nice report :)

**Comment by LanceMcCarthy** [comment]
Whoa! I thought it has been auto updating for the past couple years,lolz.

Thanks for the heads up,ill take care of that and keep an eye on things. Will come back here with more details if it persists, so you can close the Issue.

**Comment by LanceMcCarthy** [comment]
Quick follow-up, I figured out why I was still running an old version... I actually have _both_ new and old exes after the update:

`/Readarr/bin` folder has the .919 version
`/Readarr/bin/bin` has the latest .1248 version

So, I'm guessing each update was being installed to the nested bin folder? when my batch scripts launch and kill the process, it's always using the Readarr/bin folder's assemblies.

My quesiton for you then is: the app data is stored in a different location, correct? I could technically just replace everything in /bin with the content of bin/bin because it's only the compiled app that lives there?



P.S. Regarding the branch selection, I chose 'master' from a predefined drop down of available branches (it was the default). IIRC, that comes from the issue template? That might explain why you might be getting some folks stating they're using master.

**Comment by bakerboy448** [comment]
So far the only reports of the Bin issue has been if users reinstall readarr and the installer doesn't properly handle some recent changes.

Did you reinstall readarr?  Otherwise finding the update log from the funky update would useful.


As far as branch it's expected the user picks the branch their running 😬

**Comment by LanceMcCarthy** [comment]
I manually fixed the assemblies problem (it's not uncommon to see when file path is incorrect).

 In any case, it looks like the issue is no longer present in 1248. Write this up to "_user was in the woods for 12 months_" RTFM error :)

**Comment by bakerboy448** [comment]
How'd you update to 1248?

fresh installer from the website or in-app upgrade?  The former is a known issue.

If the latter can you share the updater logs from that update?

**Comment by LanceMcCarthy** [comment]
Heres exactly what i did:

1. I killed all existing processes of the app that was running 
2.  Renamed /bin to /bin_old
3. Moved the nested /bin/bin up one folder (so that it will be the new /bin)
4. Started the new .NET 6 exe

Basically, i just moved the correct bin up one level and got rid of the old one.

Edit
To answer your question..yes, i ran the installer, but that got installed to /bin/bin (instead of replacing the content of /bin). Im confient enough to say theres a bug in the updater not replacing /bin if it is current running.
